feat: add support for MSW 2.x

valendres opened the pull request:
This PR adds support for MSW 2.x, which is a breaking change. Consumers will be required to upgrade to MSW 2.x when installing playwright-msw v3. https://mswjs.io/docs/migrations/1.x-to-2.x

kettanaito commented on packages/example-vue/src/main.ts:
```diff
@@ -11,7 +11,7 @@ async function prepare() {
     return worker.start({}).then(() => {
       console.groupCollapsed('[MSW] Loaded with handlers 🎉');
-      worker.printHandlers();
+      worker.listHandlers();
```
Note that this will return an array of handlers _but you still need to print it_ to observe them. Right now, it just returns the list into the void. 

Consider adding a `console.log` here.

```js
const handlers = worker.listHandlers()
handlers.forEach(handler => console.log(handler.info.header))
```

the-ult replied:
Or mayby even the `console.info()`

valendres replied:
Ah, thanks for clarifying that. If I were better at reading, I might have noticed this in the migration guide 😅

kettanaito commented:
I may be wrong about this but it looks like you need to git-ignore the `.yarn` directory.

the-ult commented:
> I may be wrong about this but it looks like you need to git-ignore the `.yarn` directory.

Was thinking the same thing

the-ult commented on packages/example-vue/src/mocks/handler.ts:
```diff
@@ -1,11 +1,16 @@
+    '/api/users',
+
+    async () => {
+      await delay(500);
```
Wouldn't it be nice(r) the use the default `delay()` which results in random (realistic) delays?

https://mswjs.io/docs/api/delay/#implicit-delay

```ts
// Await a random realistic server response time.
await delay()
```

fostar82
--


<br class="Apple-interchange-newline">

valendres replied:
Agreed, a random delay would be a lot nicer!

the-ult commented on packages/example/src/mocks/handlers/documents.ts:
```diff
@@ -1,44 +1,40 @@
-        )
-      )
+    async ({ params }) => {
+      await delay(50);
```
@kettanaito would you recommend the usage of `delay(x)` in all/most handlers, or only occasionally?

I would think this is something you'd only want to use occasionally and. if so.. use the _realistic_ `delay()` ??

kettanaito replied:
Hard to recommend something, I'm not much aware of the intention here. 

Delaying the response must be a user's choice. I'd expect the user of this library to use `delay()` in whichever handlers they provide (if that's how it's architectured). 

You can use a realistic random delay as well. Can you tell me more about the intention here? Why is delay even needed?

valendres replied:
Nice, I didn't realize that we could call `delay()` without args to get a random amount of delay. This is awesome!

the-ult commented on packages/example/src/mocks/handlers/search.ts:
```diff
@@ -68,6 +68,9 @@ export default [
-    return response(context.status(200), context.json(searchResults));
+
+    return HttpResponse.json(searchResults, {
+      status: 200,
```
I believe the 'extra' `status: 200` is not needed here, since it is the default return status ??

valendres replied:
Good call, I didn't realize this initially. I meant to go back and clean this up, but was super pressed for time. Thank for pointing this out!

the-ult commented on packages/example/tests/playwright/specs/query-parameters.spec.ts:
```diff
@@ -51,29 +58,43 @@ test.describe.parallel('query parameters', () => {
+      http.get(
+        '/api/search?q=ignoredQuery',
+        () =>
+          new Response(
```
Is there a reason why you sometimes use the `new Response` and other times the `new HttpResponse` ?

kettanaito replied:
Unless you are mocking `Set-Cookie` response header, you should be fine using standard `Response` everywhere. But for compatibility's sake, I recommend using `HttpResponse`. This simply results in a more future-proof solution in case we decide for `HttpResponse` to handle more things in the future. 

Also, here's the [`HttpResponse` vs `Response` explanation](https://mswjs.io/docs/api/http-response#why-not-native-response).

valendres replied:
I don't have a good reason for doing this. 

Initially, I was in favor of using the standard `Response` class in favor of using native functionality. However the convenience of `HttpResponse.json` was too convenient to not use. 

I intended on circling back and using HttpResponse for all instances, but I didn't get around to it. Thankfully, @mamidenn has submitted a PR to address this :)

the-ult commented on packages/playwright-msw/README.md:
```diff
@@ -18,7 +18,9 @@
 ## Announcement
 
-Version 2 introduced a minor breaking change to the [createWorkerFixture](#createworkerfixture) function. To upgrade, please refer to the [V2 migration guide](#v2-migration-guide).
+As of version 3, `playwright-msw` has been updated to utilize MSW 2.x. In order to utilize this the latest version of `playwright-msw`, you must update your tests to use the new syntax. Please follow MSW's amazing [Migration Guide](https://mswjs.io/docs/migrations/1.x-to-2.x). Apart from updating your tests to use MSW's new syntax, the API for `playwright-msw` itself remains unchanged.
```
in order to utilize ~this~ the latest version of ...

the-ult commented on packages/playwright-msw/src/utils.ts:
```diff
@@ -66,8 +67,60 @@ export const convertMswPathToPlaywrightUrl = (path: Path): RegExp => {
+
+function isNullArray(arr: Uint8Array): boolean {
+  return (
+    arr.length === 4 && arr.every((el, idx) => el === 'null'.charCodeAt(idx))
```
Why the check on `4`? What is `4`? 
Maybe put it in a constant?

kettanaito replied:
Why not `arr.byteLength === 0`?

valendres replied:
This naive function is a way of checking if the Uint8Array is an array of char codes that represent the literal string `null`. I think this is slightly different to the behavior of `arr.byteLength === 0`.

I'm not entirely sure why this was necessary early on, but I was having issues with converting the readable stream from MSW to a buffer. If the value was "null", the route wouldn't get handled correctly. In retrospect, I suspect this was either due to me just not mocking things correctly, or just misinterpreting the issue entirely. I'm able to remove this logic and it seems to work just fine. 🤷

the-ult commented on packages/playwright-msw/src/utils.ts:
```diff
@@ -66,8 +67,60 @@ export const convertMswPathToPlaywrightUrl = (path: Path): RegExp => {
+  }
+}
+
+export const uuidv4 = () => Math.random().toString(16).slice(2);
```
Wouldn't it be better to use the node/native methods?

```ts
import { randomUUID } from 'crypto'

const uuid = randomUUID()  
``` 

https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID

kettanaito replied:
Yes, 💯 yes. This is also a global module in the browser.

valendres commented:
Thanks for the awesome feedback everyone. I've been a bit pressed for time these last few days, but I'll do my best to resolve this ASAP.

hnrq commented:
Awesome! Looking forward to this!

valendres commented:
> I may be wrong about this but it looks like you need to git-ignore the `.yarn` directory.

@the-ult @kettanaito this project is using yarn berry with zero-installs. With this feature enabled, it's intended that `.yarn/cache` is not git ignored. While it makes git clones and branch switching a lot faster, it makes code review a lot more challenging and creates some very understandable confusion. I'm going to disable this feature and rebase the PR to get rid of the unnecessarily large change set.

valendres commented:
Hey everyone,

Just a quick update on v2 support: my goal for tonight is to finalize and merge the PR, followed by a release. 🚀

I want to extend a special thanks to @the-ult and @kettanaito for your thorough reviews, and a huge shoutout to @mamidenn for proactively addressing all the feedback. Your contributions are invaluable and greatly appreciated.

I apologize for the recent delay. Between preparing for a house move, last-minute renovations for sale of the current house, and attending AWS re:INVENT, time has been particularly tight. However, now that re:INVENT has concluded and my preparations are nearly complete, I’m excited to refocus on the library.

Thank you for the patience :)

the-ult commented:
@valendres thanks you for your awesome work. Really happy with the update, can't wait to use it. 👍🏼 

These are busy times, so super happy you were able time make some time

the-ult commented on packages/example-vue/package.json:
```diff
@@ -18,8 +18,8 @@
-    "@vitejs/plugin-vue": "^4.0.0",
-    "msw": "1.0.0",
+    "@vitejs/plugin-vue": "^4.5.0",
+    "msw": "2.0.8",
```
Maybe you can do 1 more update -> `2.0.10` before merging

valendres replied:
Great idea, unfortunately I had already merged this PR. I've updated msw to `2.0.10` in a separate commit though: https://github.com/valendres/playwright-msw/commit/0740611f0973b4782784c826244c96938cb651d9

the-ult commented on packages/example/package.json:
```diff
@@ -11,23 +11,23 @@
-    "graphql": "^16.6.0",
-    "msw": "1.0.0",
+    "graphql": "^16.8.1",
+    "msw": "2.0.8",
```
latest version: 2.0.10

valendres replied:
See https://github.com/valendres/playwright-msw/pull/77#discussion_r1421292356
